feature-pitches/002-file-based-workflows/002-file-based-workflows.md

```diff
@@ -0,0 +1,106 @@
+
+
+## 🐰 Rabbit Holes
+
+### Venturing into the Metadata Realm
+
+This pitch will require a substantial effort to make the metadata all play nice. Without diving into implementation details, here are a few high-level requirments that the metadata must meet in order to provide a good user experience:
+
+* The workflows defined in files should be fully represented in the Renku KG metadata.
+* However, the workflow file itself should be the single source of truth for the workflow definition (not the KG metadata).
+* File based workflows can only be edited via the file. These workflows cannot be edited via the CLI or (in the future) somewhere in the RenkuLab UI.
+* In terms of versioning a file-based workflow, the metadata should contain metadata about the file-based workflow at each execution. There is no need to maintain a history of the file-based workflow at commits where the workflow was modified but not executed.
+* However, the KG must be aware of the latest state of the workflow file (even un-executed or not committed), because other parts of RenkuLab will be pulling information about the workflow via the KG (for example, the workflow browser will query the KG for the file-based workflow definition, and this needs to be up to date with the current state of the file so as not to confuse the user, even if the workflow file has not been executed since it was last modified).
+* It has been considered that there might be a need for a command to force-save the file-based workflow to the Renku metadata without executing it. For example, this might be helpful if the user modifies the description of the workflow, and wants to save this information to the KG without running their whole long-running workflow. However, the idea that there is a "save" beyond "file save" is a ugly area we'd like to avoid forcing upon the user if possible, so let's assume this is not needed. But the implementtion team should keep an eye out for other corner cases that might require such a command. 
+
+
+### Misc Rabbit Holes
+
+**How do you handle when a workflow step generates many files?**
+
+As is the state of the current workflow functionality, inputs and outputs may be directories, rather than files, to support multiple files. However we don’t support any kind of wildcard syntax.
+
+**How does this feature mesh with the current workflow parameter file functionality?**
+
+Currently, the user has the option of creating a file specifying workflow parameters, and passing this in to `renku workflow execute` to override the default parameters. In the future, the combination of this parameter file and the workflow definition file could develop into a sort of base workflow + several “variation” files defining a set of related workflow experiments. But work into this direction is _out of scope_ of this pitch. 
```

Comment:
I don't really think it needs to be addressed - `renku workflow execute` will continue to work even for file-based workflows (I assume) because once they are in the renku KG they will be accessible for re-execution.

Reply:
yes, that's the idea. and they can be used like normal Plans.

Reply:
@rokroskar FYI One other thing Ralf and I discussed last week is that it makes more sense conceptually for file-based workflows to be run via `renku run` than `renku workflow execute`. This is because `renku workflow execute` executes existing workflows from metadata, while `renku run` runs a user command. Plus, I think users would be much happier with the shorter syntax, too! I think it can be discussed during Build whether we want to support running the file-based workflow from both commands, or just `renku run`, etc.

Reply:
Yes definitely, it's how I understand it too. But you'll have something like this (which I think is nice!):

```
$ cat workflow.yml

name: my-workflow
step1: 
  name: do_stuff
  parameters:
    x: 1
    y: 2
step2: 
  name: do_more_stuff
  ...

$ renku run workflow.yml
# runs the workflow

$ renku workflow ls
<id> my-workflow ...

$ renku workflow execute --set step1.x=10 my-workflow
# runs my-workflow, as defined in workflow.yml and sets x=10 in step1
```

Reply:
Yes, seems like a nice freebie 🤷‍♀️